### PR TITLE
Replace #if DEBUG with conditional methods

### PR DESCRIPTION
**Pull request by bclothier**
Based on the [proposal outlined here](https://duckduckgo.com/?q=if+you%27re+using+%23debug+you%27re+doing+it+wrong&t=ffab&atb=v175-1&ia=web), and in light of recent PRs (mostly mine :( ) adding several new trace flags and conditional compilation code, I thought that Rubberduck would benefit from having more of its code verified at compile-time regardless of which flags are in effect. 

This is first of 2 commits I intend to make on this PR. The 2nd commit will be mainly within the TypeLib API where a certain fool has added a bunch of conditional compilation flags and needs ot clean up the mess. ~.~ However, before I get into this effort, I wanted to get others' feedback if this is a change for the better.

### CODE REVIEW

File: Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeInfoVBEExtensions.cs [hunk 1]
```diff
@@ -102,19 +102,22 @@ public bool CompileComponent()
...
-                    throw new ArgumentException("Unrecognised VBE compiler error: \n" + e.ToString());
-                }
-#endif
+                ThrowOnUnrecongizedCompileError(e);
 
                 return false;
             }
         }
 
+        private static void ThrowOnUnrecongizedCompileError(Exception e)
```
Comment (Vogel612) [inline Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeInfoVBEExtensions.cs]: Missing Conditional Attribute?

File: Rubberduck.Core/UI/Command/MenuItems/CommandBars/SerializeProjectsCommandMenuItem.cs [hunk 2]
```diff
@@ -77,7 +78,12 @@ protected override void OnExecute(object parameter)
...
 
-#if TRACE_COM_SAFE
+            SerializeComSafe();
+        }
+
+        [Conditional("TRACE_COM_SAFE")]
+        private void SerializeComSafe()
+        {
             //This block must be inside a conditional compilation block because the Serialize method 
             //called is conditionally compiled and available only if the compilation constant TRACE_COM_SAFE is set.
```
Comment (retailcoder) [inline Rubberduck.Core/UI/Command/MenuItems/CommandBars/SerializeProjectsCommandMenuItem.cs]: still relevant?

File: Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeInfoVBEExtensions.cs (hunk 1 above)
Comment (bclothier) [reply Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeInfoVBEExtensions.cs]: Yes, will fix.

File: Rubberduck.Core/UI/Command/MenuItems/CommandBars/SerializeProjectsCommandMenuItem.cs (hunk 2 above)
Comment (bclothier) [reply Rubberduck.Core/UI/Command/MenuItems/CommandBars/SerializeProjectsCommandMenuItem.cs]: Yes. Though it has moved out to the `Conditional("TRACE_COM_SAFE")` and the `Serialize` is no longer conditionally compiled, a build without the flag will be a no-op:

https://github.com/rubberduck-vba/Rubberduck/pull/5182/files#diff-3b3c40ef6fbf25159c4baf172e7d3ca8R33-R39

(GH doesn't want to inline the code. 😞 )

**Comment by bclothier** [comment]
The PR is now complete with its commits. However, I still need to do some live testing to verify I haven't broken something that's not a compile time error.

**Comment by bclothier** [comment]
Due to a reading comprehension failure, multiple `Conditional` attributes are not `AND`'d but rather `OR`'d. Consequently, this:

```
[Conditional("DEBUG")]
[Conditional("SOME_OTHER_FLAG")]
```

will not work the way we want it. For now, I'm going to remove the `DEBUG` flag and leave in the other flag, which is more unlikely that it'll be set, which ensures that it won't be enabled by mistake when running a debug build. 

The downside is that if the flag is left on the release build by accident, those methods will be compiled into the release build. Open to suggestions if this is a concern.

File: Rubberduck.Parsing/Common/DisabledAttribute.cs [hunk 3]
```diff
@@ -2,6 +2,12 @@
 
 namespace Rubberduck.Parsing.Common
 {
+    // Conditional cannot be an expression, so we use combined #if and an
+    // non-existent flag to ensure that the code is always checked at the
+    // compile time but omitted in non-debug builds.
+#if !DEBUG
+    [System.Diagnostics.Conditional("I_WILL_NEVER_EXIST_CRY_MY_BELOVED_DUCK")]
```
Comment (retailcoder) [inline Rubberduck.Parsing/Common/DisabledAttribute.cs]: LOL

File: Rubberduck.VBEEditor/Events/VBENativeServices.cs [hunk 4]
```diff
@@ -308,5 +294,21 @@ private static string ToClassName(this IntPtr hwnd)
             User32.GetClassName(hwnd, name, name.Capacity);
             return name.ToString();
         }
+
+        [Conditional("THIRSTY_DUCK")]
+        [Conditional("THIRSTY_DUCK_EVT")]
```
Comment (retailcoder) [inline Rubberduck.VBEEditor/Events/VBENativeServices.cs]: Does that make it `&&` or `||`?

File: Rubberduck.VBEEditor/WindowsApi/SubclassingWindow.cs [hunk 5]
```diff
@@ -86,7 +86,19 @@ public virtual int SubClassProc(IntPtr hWnd, IntPtr msg, IntPtr wParam, IntPtr l
...
+
+            if ((uint) msg == (uint) WM.DESTROY)
+            {
+                Dispose();
+            }
+            return DefSubclassProc(hWnd, msg, wParam, lParam);
+        }
+
+        [Conditional("THIRSTY_DUCK")]
+        [Conditional("THIRSTY_DUCK_WM")]
```
Comment (retailcoder) [inline Rubberduck.VBEEditor/WindowsApi/SubclassingWindow.cs]: is `[Conditional("DEBUG")]` missing?

File: Rubberduck.VBEEditor/Events/VBENativeServices.cs (hunk 4 above)
Comment (bclothier) [reply Rubberduck.VBEEditor/Events/VBENativeServices.cs]: That's a `||` - `THIRSTY_DUCK_EVT` or `THIRSTY_DUCK_WM` gives you a subset of message pump from different parts; `THRISTY_DUCK` gives you the both.

File: Rubberduck.Parsing/VBA/ParseCoordinator.cs [hunk 6]
```diff
@@ -239,13 +239,10 @@ private void ExecuteCommonParseActivities(IReadOnlyCollection<QualifiedModuleNam
             _parserStateManager.SetStatusAndFireStateChanged(this, ParserState.LoadingReference, token);
             token.ThrowIfCancellationRequested();
 
-            //TODO: Remove the conditional compilation after loading from typelibs actually works.
-#if LOAD_USER_COM_PROJECTS
```
Comment (MDoerner) [inline Rubberduck.Parsing/VBA/ParseCoordinator.cs]: This change is not mentioned in the PR comments and really should have been its own PR as it activates a part that can potentially cause RD to crash the host. This is also a candidate for the cause of the appearence of #5217.

**Pull request merged**